README
= RModBus

*RModBus* - free implementation of protocol ModBus.

== Feature

* Client ModBus-TCP
* Support functions:
  * 01 (0x01) Read Coils
  * 02 (0x02) Read Discrete Inputs
  * 03 (0x03) Read Holding Registers
  * 04 (0x04) Read Input Registers 
  * 05 (0x05) Write Single Coil
  * 06 (0x06) Write Single Register
  * 15 (0x0F) Write Multiple Coils
  * 16 (0x10) Write Multiple registers

== Installation

You can install RModBus with the following command

$ rake install

from its distribution directory

== Gem installation

Download and install RModBus with the following

$ gem sources -a http://gems.github.com
$ gem install -r sanderjd-rmodbus

== CSM

You can checkout source code from the GIT repositry 

$ git clone git://github.com/sanderjd/rmodbus.git 

== Reference 

RModBus project: http://rubyforge.org/projects/rmodbus

ModBus community: http://www.modbus-ida.org



 

 


CHANGES
2009-16-01 Release-0.2.0
- remove obsolete extension for serial communication - now this project only
  claims to support tcp, which is all it ever really did support
2009-16-01 Release-0.1.3
- fix bug in tests relating to closing sockets
2009-16-01 Release-0.1.2
- fork into git
- fixed bug in function parse_write_multiple_registers in tcp_server
- deprecated read_discret_inputs method due to misspelling - is now
  read_discrete_inputs
2008-25-01 Release-0.1.1
- fixed bug in function "read discrete inputs (0x2)" of ModBus client


test/tcp_server_spec.rb
require 'rmodbus'

describe TCPServer do
  
  before do
    @server = ModBus::TCPServer.new(8502,1)
    @server.coils = [1,0,1,1]
    @server.discret_inputs = [1,1,0,0]
    @server.holding_registers = [1,2,3,4]
    @server.input_registers = [1,2,3,4]
    @server.start
    @client = ModBus::TCPClient.new('127.0.0.1', 8502, 1)
  end

  it "should silent if UID has mismatched" do
    @client.close
    client = ModBus::TCPClient.new('127.0.0.1', 8502, 2)
    begin
      client.read_coils(1,3)
    rescue ModBus::Errors::ModBusException => ex
      ex.message.should == "Server did not respond"
    end
  end

  it "should silent if protocol identifer has mismatched" do
    socket = TCPSocket.new('127.0.0.1', 8502)
    begin
      socket.write "\0\0\1\0\0\6\1"
    rescue ModBus::Errors::ModBusException => ex
      ex.message.should == "Server did not respond"
    rescue
      # Fails if it is whiny about anything else
      assert false
    ensure
      socket.close
    end
  end

  it "should send exception if function not supported" do
    begin 
      @client.query('0x43') 
    rescue ModBus::Errors::IllegalFunction => ex
      ex.message.should == "The function code received in the query is not an allowable action for the server"  
    end 
  end

  it "should send exception if quanity of out more 0x7d" do
    begin
      @client.read_coils(0, 0x7e)
    rescue ModBus::Errors::IllegalDataValue => ex
      ex.message.should == "A value contained in the query data field is not an allowable value for server"
    end
  end

  it "should send exception if addr not valid" do
    begin
      @client.read_coils(2, 8)
    rescue ModBus::Errors::IllegalDataAddress => ex
      ex.message.should == "The data address received in the query is not an allowable address for the server"
    end
  end

  it "should supported function 'read coils'" do
    @client.read_coils(0,3).should == @server.coils[0,3]
  end

  it "should supported function 'read discrete inputs'" do
    @client.read_discrete_inputs(1,3).should == @server.discret_inputs[1,3]
  end

  it "should supported function 'read holding registers'" do
    @client.read_holding_registers(0,3).should == @server.holding_registers[0,3]
  end

  it "should supported function 'read input registers'" do
    @client.read_input_registers(2,2).should == @server.input_registers[2,2]
  end

  it "should supported function 'write single coil'" do
    @server.coils[3] = 0
    @client.write_single_coil(3,1)
    @server.coils[3].should == 1
  end

  it "should supported function 'write single register'" do
    @server.holding_registers[3] = 25
    @client.write_single_register(3,35)
    @server.holding_registers[3].should == 35
  end

  it "should supported function 'write multiple coils'" do
    @server.coils = [1,1,1,0, 0,0,0,0, 0,0,0,0, 0,1,1,1]
    @client.write_multiple_coils(3, [1, 0,1,0,1, 0,1,0,1])
    @server.coils.should == [1,1,1,1, 0,1,0,1, 0,1,0,1, 0,1,1,1]
  end

  it "should supported function 'write multiple registers'" do
    @server.holding_registers = [1,2,3,4,5,6,7,8,9]
    @client.write_multiple_registers(3,[1,2,3,4,5])
    @server.holding_registers.should == [1,2,3,1,2,3,4,5,9]
  end
  
  after do
    @client.close if @client
    @server.stop if @server
  end
end


lib/rmodbus/client.rb
# RModBus - free implementation of ModBus protocol on Ruby.
#
# Copyright (C) 2008  Timin Aleksey
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

require 'rmodbus/exceptions'

class String

  def to_int16
    self[0]*256 + self[1]
  end

  def to_array_int16
    array_int16 = []
    i = 0
    while(i < self.size) do
      array_int16 << self[i]*256 + self[i+1]
      i += 2
    end
    array_int16
  end

  def to_array_bit
    array_bit = []
    self.each_byte do |byte|
      mask = 0x01 
      8.times {
        unless (mask & byte) == 0
          array_bit << 1
        else
          array_bit << 0
        end
        mask = mask << 1
      }
    end
    array_bit
  end

end

class Integer
  def to_bytes
    (self >> 8).chr + (self & 0xff).chr 
  end
end

class Array

  def to_ints16
    s = ""
    self.each do |int16|
      s << int16.to_bytes
    end
    s
  end

  def bits_to_bytes
    int16 = 0
    s = ""
    mask = 0x01

    self.each do |bit| 
      int16 |= mask if bit > 0
      mask <<= 1
      if mask  == 0x100
        mask = 0x01
        s << int16.chr
        int16 = 0
      end
    end
    s << int16.chr unless mask == 0x01
  end

end

module ModBus

  class Client
  
    include Errors

    # Number of times to retry on connection and read timeouts
    CONNECTION_RETRIES = 10
    READ_RETRIES = 10

    # Read value *ncoils* coils starting with *addr*
    #
    # Return array of their values
    def read_coils(addr, ncoils)
      query("\x1" + addr.to_bytes + ncoils.to_bytes).to_array_bit[0..ncoils-1]
    end

    # Read value *ncoils* discrete inputs starting with *addr*
    #
    # Return array of their values
    def read_discrete_inputs(addr, ncoils)
      query("\x2" + addr.to_bytes + ncoils.to_bytes).to_array_bit[0..ncoils-1]
    end

    # Deprecated version of read_discrete_inputs
    def read_discret_inputs(addr, ncoils)
      warn "[DEPRECATION] `read_discret_inputs` is deprecated.  Please use `read_discrete_inputs` instead."
      read_discrete_inputs(addr, ncoils)
    end

    # Read value *nreg* holding registers starting with *addr*
    #
    # Return array of their values
    def read_holding_registers(addr, nreg) 
      query("\x3" + addr.to_bytes + nreg.to_bytes).to_array_int16
    end

    # Read value *nreg* input registers starting with *addr*
    #
    # Return array of their values
    def read_input_registers(addr, nreg)
      query("\x4" + addr.to_bytes + nreg.to_bytes).to_array_int16
    end

    # Write *val* in *addr* coil 
    #
    # if *val* lager 0 write 1
    #
    # Return self
    def write_single_coil(addr, val)
      if val == 0
        query("\x5" + addr.to_bytes + 0.to_bytes)
      else
        query("\x5" + addr.to_bytes + 0xff00.to_bytes) 
      end
      self
    end

    # Write *val* in *addr* register
    #
    # Return self
    def write_single_register(addr, val)
      query("\x6" + addr.to_bytes + val.to_bytes)
      self
    end

    # Write *val* in coils starting with *addr*
    #
    # *val* it is array of bits
    #
    # Return self
    def write_multiple_coils(addr, val)
      nbyte = ((val.size-1) >> 3) + 1
      sum = 0
      (val.size - 1).downto(0) do |i|
        sum = sum << 1
        sum |= 1 if val[i] > 0
      end
   
      s_val = ""
      nbyte.times do
        s_val << (sum & 0xff).chr
        sum >>= 8 
      end

      query("\xf" + addr.to_bytes + val.size.to_bytes + nbyte.chr + s_val)
      self
    end

    # Write *val* in registers starting with *addr*
    #
    # *val* it is array of integer
    #
    # Return self
    def write_multiple_registers(addr, val)
      s_val = ""
      val.each do |reg|
        s_val << reg.to_bytes
      end

      query("\x10" + addr.to_bytes + val.size.to_bytes + (val.size * 2).chr + s_val)
      self
    end

    # Write *current value & and_mask | or mask in *addr* register
    #
    # Return self
    def mask_write_register(addr, and_mask, or_mask)
      query("\x16" + addr.to_bytes + and_mask.to_bytes + or_mask.to_bytes)
      self  
    end

    def query(pdu)    
      send_pdu(pdu)

      tried = 0
      begin
        timeout(1, ModBusTimeout) do
          pdu = read_pdu
        end
      rescue ModBusTimeout => err
        tried += 1
        retry unless tried >= READ_RETRIES
        raise ModBusTimeout.new, 'Timed out during read attempt'
      end

      if pdu[0].to_i >= 0x80
        case pdu[1].to_i
          when 1
            raise IllegalFunction.new, "The function code received in the query is not an allowable action for the server"  
          when 2
            raise IllegalDataAddress.new, "The data address received in the query is not an allowable address for the server"
          when 3
            raise IllegalDataValue.new, "A value contained in the query data field is not an allowable value for server"
          when 4
            raise SlaveDeviceFailure.new, "An unrecoverable error occurred while the server was attempting to perform the requested action"
          when 5
            raise Acknowledge.new, "The server has accepted the request and is processing it, but a long duration of time will be required to do so"
          when 6
            raise SlaveDeviceBus.new, "The server is engaged in processing a long duration program command"
          when 8
            raise MemoryParityError.new, "The extended file area failed to pass a consistency check"
          else
            raise ModBusException.new, "Unknown error"
        end
      end
      pdu[2..-1]
    end

    protected
    def send_pdu(pdu)
    end

    def read_pdu
    end

    def close
    end

  end

end


lib/rmodbus/exceptions.rb
# RModBus - free implementation of ModBus protocol on Ruby.
#
# Copyright (C) 2008  Timin Aleksey
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
module ModBus

  module Errors

    class ModBusException < Exception
    end

    class ModBusTimeout < ModBusException
    end

    class IllegalFunction < ModBusException
    end

    class IllegalDataAddress < ModBusException
    end

    class IllegalDataValue < ModBusException
    end

    class SlaveDeviceFailure < ModBusException
    end

    class Acknowledge < ModBusException
    end

    class SlaveDeviceBus < ModBusException
    end

    class MemoryParityError < ModBusException
    end

  end

end


lib/rmodbus/tcp_client.rb
# RModBus - free implementation of ModBus protocol on Ruby.
#
# Copyright (C) 2008  Timin Aleksey
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
require 'socket'
require 'timeout'
require 'rmodbus/client'
require 'rmodbus/exceptions'

module ModBus

  # Implementation clients(master) ModBusTCP
  class TCPClient < Client

    include Timeout

    @@transaction = 0

    # Connect with a ModBus server
    def initialize(ipaddr, port = 502, slaveaddr = 1)
      tried = 0
      begin
        timeout(1, ModBusTimeout) do
          @sock = TCPSocket.new(ipaddr, port)
        end
      rescue ModBusTimeout => err
        tried += 1
        retry unless tried >= CONNECTION_RETRIES
        raise ModBusTimeout.new, 'Timed out attempting to create connection'
      end

      @slave = slaveaddr
    end

    def close
      @sock.close unless @sock.closed?
    end
 
    def self.transaction 
      @@transaction
    end

    private
    def send_pdu(pdu)   
      @@transaction += 1 
      @sock.write @@transaction.to_bytes + "\0\0" + (pdu.size + 1).to_bytes + @slave.chr + pdu
    end

    def read_pdu     
      header = @sock.read(7)            
      if header
        tin = header[0,2].to_int16
        raise Errors::ModBusException.new("Transaction number mismatch") unless tin == @@transaction
        len = header[4,2].to_int16       
        @sock.read(len-1)               
      else
        raise Errors::ModBusException.new("Server did not respond")
      end
    end

  end

end


lib/rmodbus/tcp_server.rb
# RModBus - free implementation of ModBus protocol on Ruby.
#
# Copyright (C) 2008  Timin Aleksey
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
require 'rmodbus/exceptions'
require 'gserver'


module ModBus
	
	class TCPServer < GServer
    
    attr_accessor :coils, :discret_inputs, :holding_registers, :input_registers

    @@funcs = [1,2,3,4,5,6,15,16]
    
    def initialize(port = 502, uid = 1)
      @coils = []
      @discret_inputs = []
      @holding_registers =[]
      @input_registers = []
      @uid = uid
      super(port)
    end

    def serve(io)
      req = io.read(7)
      if req[2,2] != "\x00\x00" or req[6].to_i != @uid
        io.close
        return
      end

      tr = req[0,2]
      len = req[4,2].to_int16
      req = io.read(len - 1)
      func = req[0].to_i

      unless @@funcs.include?(func)
        param = { :err => 1 }
      end
      
      case func
        when 1
          param = parse_read_func(req, @coils)
          if param[:err] == 0
            val = @coils[param[:addr],param[:quant]].bits_to_bytes
            res = func.chr + val.size.chr + val
          end
        when 2
          param = parse_read_func(req, @discret_inputs)
          if param[:err] == 0
            val = @discret_inputs[param[:addr],param[:quant]].bits_to_bytes
            res = func.chr + val.size.chr + val
          end
        when 3
          param = parse_read_func(req, @holding_registers)
          if param[:err] == 0
            res = func.chr + (param[:quant] * 2).chr + @holding_registers[param[:addr],param[:quant]].to_ints16
          end
        when 4
          param = parse_read_func(req, @input_registers)
          if param[:err] == 0
            res = func.chr + (param[:quant] * 2).chr + @input_registers[param[:addr],param[:quant]].to_ints16
          end
        when 5 
          param = parse_write_coil_func(req)
          if param[:err] == 0
            @coils[param[:addr]] = param[:val]
            res = func.chr + req
          end
        when 6
          param = parse_write_register_func(req)
          if param[:err] == 0
            @holding_registers[param[:addr]] = param[:val]
            res = func.chr + req
          end
        when 15
          param = parse_write_multiple_coils_func(req)
          if param[:err] == 0
            @coils[param[:addr],param[:quant]] = param[:val][0,param[:quant]]
            res = func.chr + req
          end
        when 16
          param = parse_write_multiple_registers_func(req)
          if param[:err] == 0
            @holding_registers[param[:addr],param[:quant]] = param[:val][0,param[:quant]]
            res = func.chr + req
          end
      end
      if param[:err] ==  0
        io.write tr + "\0\0" + (res.size + 1).to_bytes + @uid.chr + res
      else
        io.write tr + "\0\0\0\3" + @uid.chr + (func | 0x80).chr + param[:err].chr
      end 
    end

    private

    def parse_read_func(req, field)
      quant = req[3,2].to_int16

      return { :err => 3} unless quant <= 0x7d
        
      addr = req[1,2].to_int16
      return { :err => 2 } unless addr + quant <= field.size
        
      return { :err => 0, :quant => quant, :addr => addr }    
    end

    def parse_write_coil_func(req)
      addr = req[1,2].to_int16
      return { :err => 2 } unless addr <= @coils.size

      val = req[3,2].to_int16
      return { :err => 3 } unless val == 0 or val == 0xff00
      
      val = 1 if val == 0xff00
      return { :err => 0, :addr => addr, :val => val }  
    end

    def parse_write_register_func(req)
      addr = req[1,2].to_int16
      return { :err => 2 } unless addr <= @coils.size

      val = req[3,2].to_int16

      return { :err => 0, :addr => addr, :val => val }  
	  end

    def parse_write_multiple_coils_func(req)
      param = parse_read_func(req, @coils)

      if param[:err] == 0
        param = {:err => 0, :addr => param[:addr], :quant => param[:quant], :val => req[6,param[:quant]].to_array_bit }
      end
      param
    end

    def parse_write_multiple_registers_func(req)
      param = parse_read_func(req, @holding_registers)

      if param[:err] == 0
        # Each val is 2 bytes, so multiply by two to get number of bytes
        bytes_for_vals = param[:quant] * 2
        param = {:err => 0, :addr => param[:addr], :quant => param[:quant], :val => req[6,bytes_for_vals].to_array_int16 }
      end
      param
    end

	end
end
